File: tutor-claude/jpa-tutor/src/main/java/com/jpatutor/ch07_proxy_fetch/Order.java
package com.jpatutor.ch07_proxy_fetch;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.OneToMany;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

/**
 * [7장] Order는 OrderItem들의 "생명주기를 통째로 책임지는" 애그리거트 루트다.
 *
 * - cascade = CascadeType.ALL: Order를 persist/remove 하면 그 상태 변화가 자식(OrderItem)에게
 *   전이(cascade)된다. 즉 order만 persist해도 컬렉션에 담긴 orderItem들이 자동으로 함께 저장된다.
 *   주의: cascade는 "연관관계의 주인 여부"와 무관한 별개의 기능이고, 오직 Order를 통해서만
 *   OrderItem을 저장/삭제하는 소유 관계(private ownership)일 때만 사용해야 한다. 여러 부모가
 *   자식을 공유하는 경우에 cascade ALL을 쓰면 다른 부모가 삭제될 때 자식까지 같이 삭제되는
 *   사고로 이어질 수 있다.
 * - orphanRemoval = true: 컬렉션(orderItems)에서 자식을 제거하면(items.remove(item)),
 *   그 자식은 "고아(orphan)"가 된 것으로 간주되어 DB에서도 실제로 DELETE된다.
 *   cascade와 orphanRemoval을 함께 쓰면, Order 하나만 갖고도 OrderItem 전체 생명주기를
 *   완전히 제어할 수 있다 (DDD의 애그리거트 루트 패턴과 정확히 맞아떨어진다).
 * - fetch = LAZY: 컬렉션 연관관계는 항상 LAZY로 설정한다 (@OneToMany의 기본값도 원래 LAZY지만
 *   명시적으로 적어서 의도를 분명히 한다).
 */
@Entity
@Table(name = \"ch07_order\")
@Getter
@NoArgsConstructor
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String orderer;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<OrderItem> orderItems = new ArrayList<>();

    public Order(String orderer) {
        this.orderer = orderer;
    }

    public void addItem(OrderItem item) {
        orderItems.add(item);
        item.assignOrder(this);
    }

    public void removeItem(OrderItem item) {
        // 이 한 줄만으로 DB에서 해당 order_item 로우가 DELETE된다 (orphanRemoval 덕분에).
        orderItems.remove(item);
    }
}

File: tutor-claude/springboot-tutor/src/main/kotlin/be/com/springbootclaude/ddd/order/domain/model/Order.kt
package be.com.springbootclaude.ddd.order.domain.model

import be.com.springbootclaude.ddd.order.domain.event.OrderCancelled
import be.com.springbootclaude.ddd.order.domain.event.OrderPlaced
import jakarta.persistence.*
import java.time.LocalDateTime

/**
 * Aggregate Root: Order (주문)
 *
 * ★ DDD에서 가장 중요한 개념: Aggregate Root ★
 *
 * Aggregate란?
 * - 관련된 Entity와 Value Object의 집합
 * - 데이터 변경의 단위 (트랜잭션 경계)
 * - 일관성을 보장하는 경계
 *
 * Aggregate Root란?
 * - Aggregate의 진입점이 되는 Entity
 * - 외부에서는 오직 Root를 통해서만 Aggregate 내부에 접근
 * - 불변식(Invariant)을 보장할 책임
 *
 * 왜 Order가 Aggregate Root인가?
 * 1. OrderItem은 Order 없이 의미 없음
 * 2. OrderItem의 변경은 Order의 총액에 영향
 * 3. Order가 OrderItem들의 일관성 보장
 *
 * 핵심 규칙:
 * - 외부에서 OrderItem을 직접 생성/변경하면 안 됨
 * - 모든 변경은 Order를 통해서만 (addItem, removeItem 등)
 * - Order가 비즈니스 규칙 검증
 */
@Entity
@Table(name = "orders")
class Order(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(nullable = false)
    val customerId: Long,

    @Embedded
    @AttributeOverrides(
        AttributeOverride(name = "zipCode", column = Column(name = "delivery_zip_code")),
        AttributeOverride(name = "city", column = Column(name = "delivery_city")),
        AttributeOverride(name = "street", column = Column(name = "delivery_street")),
        AttributeOverride(name = "detailAddress", column = Column(name = "delivery_detail_address"))
    )
    val deliveryAddress: Address,

    /**
     * Aggregate 경계: OrderItem은 Order에 완전히 종속
     * CascadeType.ALL: Order 저장/삭제 시 OrderItem도 함께
     * orphanRemoval: Order에서 제거된 OrderItem은 자동 삭제
     */
    @OneToMany(
        mappedBy = "order",
        cascade = [CascadeType.ALL],
        orphanRemoval = true,
        fetch = FetchType.LAZY
    )
    private val _orderItems: MutableList<OrderItem> = mutableListOf(),

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: OrderStatus = OrderStatus.PENDING,

    @Column(nullable = false)
    val orderedAt: LocalDateTime = LocalDateTime.now(),

    /**
     * Domain Events: Aggregate에서 발생한 중요한 사건
     * 다른 Bounded Context나 외부 시스템에 알릴 필요가 있는 이벤트
     */
    @Transient
    private val _domainEvents: MutableList<Any> = mutableListOf()
) {
    /**
     * Aggregate 불변식 보호: 외부에서 직접 수정 불가
     * 읽기 전용 뷰 제공
     */
    val orderItems: List<OrderItem>
        get() = _orderItems.toList()

    val domainEvents: List<Any>
        get() = _domainEvents.toList()

    init {
        require(customerId > 0) { "고객 ID는 양수여야 합니다." }
    }

    /**
     * ★ Aggregate Root의 핵심 메서드 ★
     * 외부에서는 이 메서드로만 OrderItem 추가
     * 비즈니스 규칙을 검증하고 일관성 보장
     */
    fun addItem(productInfo: ProductInfo, quantity: Int) {
        require(status == OrderStatus.PENDING) {
            "주문 대기 상태에서만 항목을 추가할 수 있습니다. 현재 상태: $status"
        }

        // 같은 상품이 이미 있으면 수량만 증가
        val existingItem = _orderItems.find { it.productInfo.productId == productInfo.productId }
        if (existingItem != null) {
            existingItem.changeQuantity(existingItem.quantity + quantity)
        } else {
            val newItem = OrderItem(
                productInfo = productInfo,
                quantity = quantity,
                order = this
            )
            _orderItems.add(newItem)
        }
    }

    /**
     * 항목 제거: Aggregate Root가 일관성 보장
     */
    fun removeItem(productId: Long) {
        require(status == OrderStatus.PENDING) {
            "주문 대기 상태에서만 항목을 제거할 수 있습니다."
        }

        _orderItems.removeIf { it.productInfo.productId == productId }

        require(_orderItems.isNotEmpty()) {
            "주문에는 최소 1개 이상의 항목이 필요합니다."
        }
    }

    /**
     * 도메인 로직: 총 주문 금액 계산
     * Money Value Object의 연산 활용
     */
    fun calculateTotalAmount(): Money {
        return _orderItems
            .map { it.calculateTotalPrice() }
            .fold(Money.ZERO) { acc, money -> acc + money }
    }

    /**
     * 도메인 로직: 주문 확정
     * 비즈니스 규칙 검증 + 상태 변경 + 이벤트 발행
     */
    fun confirm() {
        require(status == OrderStatus.PENDING) {
            "대기 중인 주문만 확정할 수 있습니다. 현재 상태: $status"
        }
        require(_orderItems.isNotEmpty()) {
            "주문 항목이 없습니다."
        }

        this.status = OrderStatus.CONFIRMED

        // Domain Event 발행
        val totalAmount = calculateTotalAmount()
        _domainEvents.add(
            OrderPlaced(
                orderId = id!!,
                customerId = customerId,
                totalAmount = totalAmount,
                itemCount = _orderItems.size
            )
        )
    }

    /**
     * 도메인 로직: 주문 취소
     * 상태 전이 규칙 검증
     */
    fun cancel(reason: String) {
        require(status.canBeCancelled()) {
            "현재 상태($status)에서는 주문을 취소할 수 없습니다."
        }

        val previousStatus = this.status
        this.status = OrderStatus.CANCELLED

        // Domain Event 발행
        _domainEvents.add(
            OrderCancelled(
                orderId = id!!,
                customerId = customerId,
                reason = reason,
                previousStatus = previousStatus
            )
        )
    }

    /**
     * 도메인 로직: 상태 변경
     * 비즈니스 규칙에 따른 전이만 허용
     */
    fun changeStatus(newStatus: OrderStatus) {
        require(status.canTransitionTo(newStatus)) {
            "주문 상태를 $status 에서 $newStatus 로 변경할 수 없습니다."
        }
        this.status = newStatus
    }

    /**
     * 도메인 이벤트 클리어 (이벤트 발행 후 호출)
     */
    fun clearDomainEvents() {
        _domainEvents.clear()
    }

    /**
     * Entity는 ID로 동등성 비교
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Order) return false
        return id != null && id == other.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}
